### ISSUE TITLE
Partitions related to a single entity type should not be executed in parallel

### ISSUE THREAD

**Issue by yrodiere** [mincong-h/gsoc-hsearch#166]
The current batch seems to mix partitions for all entity types in the same partition list, meaning that two partitions relating to a single entity type may be parallelized. See https://github.com/mincong-h/gsoc-hsearch/blob/89624d52f7b6b955bbf45d576b538b8fc93445fb/core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/PartitionMapper.java#L140

While it may be good to index different entity types in parallel, I'm not sure executing two "chunks" for the same entity type in parallel is a good idea. We may have locks on the same tables for both chunks, and writes/flushes to the same index for both chunks.
I agree the database/index probably have strategies in place to mitigate the slowdown this would create, but we might as well not push our luck...

For the record, the current implementation of `MassIndexer` in Hibernate Search never indexes two entities of the same type in parallel; see `org.hibernate.search.batchindexing.impl.BatchCoordinator.doBatchWork(BatchBackend)`.

If someone thinks the current behavior could on the contrary improve performance, maybe we should benchmark it.

Related ticket: #4, where performance of the batch indexing is discussed.

**Comment by mincong-h** [comment]
> We may have locks on the same tables for both chunks, and writes/flushes to the same index for both chunks.

Do you mean the lock of database or lock of Lucene? If you mean the lock of database, the parallel processing on the same entity type will be fine since loading data is a READ-ONLY operation. If you mean the lock of Lucene, well I think you're right...

**Comment by yrodiere** [comment]
> Do you mean the lock of database or lock of Lucene? If you mean the lock of database, the parallel processing on the same entity type will be fine since loading data is a READ-ONLY operation. If you mean the lock of Lucene, well I think you're right...

I meant both. And yeah, I guess shared locks on the tables shouldn't be that bad... Though I'd be wary when making any kind of assumptions about this stuff (I personally don't know enough).

Thinking about it, the main issue here probably isn't that we are processing the same entity type in multiple threads at the same time; it's more that while we're busy processing this entity type on multiple threads, there probably are entity types not being processed at all (I assume in most cases there will be more entity types than processing threads).
So every now and then, there are resources that we don't use at all (the indexes for these entities) while other resources are used by multiple threads (the indexes/tables for entity types processed by multiple threads), which basically means the parallelization is not very... parallel. Worst case, all threads could be busy on the same set of resources.

Please note I'm only mentioning this because I saw in #4 that performance may not be as good with this implementation as with the current `MassIndexerImpl`. This may be one of the reasons, even though I have no proof right now.

Bad news is, I looked a bit at the JSR-352 spec, and I don't see anything that would allow us to do what the current `MassIndexImpl` does:

 * we have no control over how partitions are executed (we can't have sequential subsets of partitions)
 * the `split` constructs seems completely static, with no way to define one `flow` per entity dynamically, based on batch parameters.

JBeret seems to offer a dynamic API that would allow to build our batch description dynamically (thus allowing to define one `flow` per entity), but this is not part of the specX.

So... I guess there's not much we can do right now :/

**Comment by yrodiere** [comment]
Closing as per our Hangout discussion: we expect to have much more threads than entity types, so the issues I'm describing here are not revelant.
